### PR TITLE
Predicate subgroups

### PR DESCRIPTION
**Pull request by Alizter**
This is a WIP attempt at switching to predicates for subgroups as discussed in #1387.

This should have a few advantages when it comes to formalization.

I've also implemented a basic hint database for groups, which can automate a lot of routine arguments that we had. With some more careful design we can make this a powerful and useful tool.

This should also resolve issues we were having with #1369 though I haven't been able to confirm this yet.

When it comes to actually constructing subgroups, the predicate approach is far more efficient. There is no need to construct another group and check that the inclusion map is a homomorphism, etc.

### CODE REVIEW

File: theories/Algebra/Groups/Kernel.v [hunk 1]
```diff
@@ -123,4 +91,4 @@ Proof.
     + apply equiv_path_grouphomomorphism; intro x; cbn.
       refine (ap pr1 (y:=(mon_unit; grp_homo_unit f)) _).
       apply path_ishprop.
-Defined.
+Defined.*)
```
Comment (jdchristensen) [inline theories/Algebra/Groups/Kernel.v:94]: Why is this result commented out?

File: theories/Algebra/Groups/ShortExactSequence.v [hunk 2]
```diff
@@ -46,10 +46,12 @@ Proof.
...
 
+(*
 (** A complex 0 -> A -> B is purely exact if and only if the kernel of the map A -> B is trivial. *)
 Corollary equiv_grp_isexact_kernel `{Univalence} {A B : Group} (f : A $-> B)
   : IsExact purely (@grp_homo_const grp_trivial A) f
             <~> (grp_kernel f = trivial_subgroup).
 Proof.
   exact ((equiv_kernel_isembedding f)^-1%equiv oE equiv_grp_isexact_isembedding f).
 Defined.
+*)
```
Comment (jdchristensen) [inline theories/Algebra/Groups/ShortExactSequence.v:57]: No newline at end of file.

File: theories/Algebra/Groups/Subgroup.v [hunk 3]
```diff
@@ -7,91 +7,130 @@ Generalizable Variables G H A B C N f g.
...
-Class IsSubgroup (G H : Group) := {
-  issubgroup_incl : GroupHomomorphism G H;
-  isinj_issubgroup_incl : IsInjective issubgroup_incl;
+Class IsSubgroup {G : Group} (H : G -> Type) := {
+  ishprop_issubgroup : forall x, IsHProp (H x) ;
+  issubgroup_unit : H mon_unit ;
+  issubgroup_op : forall x y, H x -> H y -> H (x * y) ;
+  issubgroup_inv : forall x, H x -> H (- x) ;
+(*   issubgroup_incl : GroupHomomorphism G H; *)
+(*   isinj_issubgroup_incl : IsInjective issubgroup_incl; *)
```
Comment (jdchristensen) [inline theories/Algebra/Groups/Subgroup.v:16]: Delete last two lines?

File: theories/Algebra/Groups/Subgroup.v [hunk 4]
```diff
@@ -7,91 +7,130 @@ Generalizable Variables G H A B C N f g.
...
 
-Definition issig_issubgroup G H : _ <~> IsSubgroup G H
+Definition issig_issubgroup {G : Group} (H : G -> Type) : _ <~> IsSubgroup H
   := ltac:(issig).
 
-(** Given two groups G and H, there is a set worth of ways H can be a subgroup of G, corresponding to the set of injections H $-> G. *)
-Global Instance ishset_issubgroup `{U : Univalence} {G H : Group}
-  : IsHSet (IsSubgroup G H).
+Global Instance ishset_issubgroup `{F : Funext} {G : Group} {H : G -> Type}
+  : IsHProp (IsSubgroup H).
```
Comment (jdchristensen) [inline theories/Algebra/Groups/Subgroup.v:34]: The new statement and the old one aren't saying the same things.  Maybe that's intended.  But then the name should change to ishprop_issubgroup.

File: theories/Basics/Tactics.v [hunk 5]
```diff
@@ -42,7 +42,6 @@ Ltac on_last_hyp tac :=
 
 Ltac destruct_one_pair :=
  match goal with
-   | [H : conj _ _ |- _] => destruct H
```
Comment (jdchristensen) [inline theories/Basics/Tactics.v:45]: Why are seemingly unrelated changes happening in this file?

**Review by jdchristensen — COMMENTED** [review]
This looks like a good start.  I didn't review the second half carefully, as the diff becomes hard to follow.  I see that there are lots of things commented out, so more work needs to be done.

**Comment by jdchristensen** [comment]
It's very impressive how much shorter some of the proofs became with the hints database.  Would it be hard to break this into two PR's, one about the change to IsSubgroup and one about the hints?

**Comment by Alizter** [comment]
Yeah, this PR wasn't quite ready for full review yet :-) I made it so that @jarlg could give me a hand with some of the commented out proofs. As you say, it might be worth splitting this into two PRs. I can probably get the hints merged before the predicate stuff.

File: theories/Basics/Tactics.v (hunk 5 above)
Comment (Alizter) [reply theories/Basics/Tactics.v:45]: I've factored this out into #1441

**Comment by Alizter** [comment]
OK so @jarlg and I went ahead and did some cleanup. This PR is now solely about switching to predicates and it makes many proofs much shorter and cleaner. There is still some more cleanup work to do however.

**Comment by Alizter** [comment]
OK so this should be in working order now. There is still some clean up left to do.

It's worth noting that the definition of being a ring ideal has been vastly simplified by using the predicate approach. In fact it also makes it more obvious how to construct various operations on ideals such as intersections. We need to of course have the operations on the underlying subgroups first. I will leave these for a later PR however.

File: theories/Algebra/AbGroups/AbelianGroup.v [hunk 6]
```diff
@@ -94,8 +89,19 @@ Proof.
   apply commutativity.
 Defined.
 
-Definition QuotientAbGroup (G : AbGroup) (H : Subgroup G)
-  : AbGroup := Build_AbGroup (QuotientGroup G H) _ _ _ _.
+Definition QuotientAbGroup (G : AbGroup) (H : Subgroup G) : AbGroup :=
+  Build_AbGroup
+    (QuotientGroup G
+      (Build_NormalSubgroup G H (isnormal_ab_subgroup G H)))
```
Comment (jdchristensen) [inline theories/Algebra/AbGroups/AbelianGroup.v:95]: It looks like it would be handy to have `QuotientGroup'` which takes `G : Group`, `H : Subgroup G` and `k : IsNormalSubgroup H`.  There are several places that could use it.

**Review by jdchristensen — COMMENTED** [review]
Nice.  One reason this ends up simpler is that to give a subgroup in predicate style, you just have to give the predicate and show closure under the operations.  In the previous style, you had to give the operations on the subgroup, prove all of the group operations, and show that the inclusion map respects the operations, and much of that was in fact redundant.

File: theories/Algebra/AbGroups/AbelianGroup.v (hunk 6 above)
Comment (Alizter) [reply theories/Algebra/AbGroups/AbelianGroup.v]: OK done.

File: theories/Algebra/Groups/QuotientGroup.v [hunk 7]
```diff
@@ -166,6 +166,10 @@ Arguments grp_quotient_map {_ _}.
 
 Notation "G / N" := (QuotientGroup G N) : group_scope.
 
+(** Rephrasing that let's you specify the normality proof *)
```
Comment (jdchristensen) [inline theories/Algebra/Groups/QuotientGroup.v:169]: Should be "lets".  Does anyone know why the "+-" button for proposing a diff isn't showing for me?

Comment (jdchristensen) [reply theories/Algebra/Groups/QuotientGroup.v:169]:
```suggestion
(** Rephrasing that lets you specify the normality proof *)
```

File: theories/Algebra/Groups/QuotientGroup.v [hunk 8]
```diff
@@ -166,6 +166,10 @@ Arguments grp_quotient_map {_ _}.
...
+Definition QuotientGroup' (G : Group) (N : Subgroup G) (H : IsNormalSubgroup N)
```
Comment (jdchristensen) [inline theories/Algebra/Groups/QuotientGroup.v:170]: test

File: theories/Algebra/Groups/QuotientGroup.v [hunk 9]
```diff
@@ -139,51 +139,53 @@ Section QuotientGroup.
...
   Defined.
 
 End QuotientGroup.
 
-Arguments grp_quotient_map {_ _ _ _}.
+Arguments grp_quotient_map {_ _}.
 
 Notation "G / N" := (QuotientGroup G N) : group_scope.
 
+(** Rephrasing that let's you specify the normality proof *)
```
Comment (jdchristensen) [inline theories/Algebra/Groups/QuotientGroup.v:169]: Should be "lets".

File: theories/Algebra/Groups/QuotientGroup.v (hunk 7 above)
Comment (jdchristensen) [inline theories/Algebra/Groups/QuotientGroup.v:169]: Should be "lets".

Comment (jdchristensen) [reply theories/Algebra/Groups/QuotientGroup.v:169]:
```suggestion
(** Rephrasing that lets you specify the normality proof *)
```

**Review by mikeshulman — APPROVED** [review]
Great!  I always thought this would be the better way to do it.

**Comment by Alizter** [comment]
OK, rebased and @jarlg did some more cleanup

**Comment by andreaslyn** [comment]
Perhaps we should do something about the large PR https://github.com/HoTT/HoTT/pull/1445 first (merge/close)?

**Comment by Alizter** [comment]
@andreaslyn I think there is still some discussion to be had there.
